scenario/state.py

```diff
@@ -643,24 +644,63 @@ def _generate_new_change_id():
 
 
 @dataclasses.dataclass(frozen=True)
-class ExecOutput(_max_posargs(0)):
+class Exec(_max_posargs(1)):
     """Mock data for simulated :meth:`ops.Container.exec` calls."""
 
+    command_prefix: Sequence[str]
     return_code: int = 0
-    """The return code of the process (0 is success)."""
+    """The return code of the process.
+
+    Use 0 to mock the process ending successfully, and other values for failure.
+    """
     stdout: str = ""
-    """Any content written to stdout by the process."""
+    """Any content written to stdout by the process.
+
+    Provide content that the real process would write to stdout, which can be
+    read by the charm.
+    """
     stderr: str = ""
-    """Any content written to stderr by the process."""
+    """Any content written to stderr by the process.
+
+    Provide content that the real process would write to stderr, which can be
+    read by the charm.
+    """
+    _stdin: Optional[str] = None
+    """stdin content the charm wrote to the process.
+
+    Cannot be provided in the input state - the output state will contain the
+    content the charm wrote to the process, to use in assertions.
+    """
 
     # change ID: used internally to keep track of mocked processes
     _change_id: int = dataclasses.field(default_factory=_generate_new_change_id)
 
+    def __post_init__(self):
+        # The command prefix can be any sequence type, and a list is tidier to
+        # write when there's only one string. However, this object needs to be
+        # hashable, so can't contain a list. We 'freeze' the sequence to a tuple
+        # to support that.
+        object.__setattr__(self, "command_prefix", tuple(self.command_prefix))
+
+        # The process is being mocked - it can't start with existing stdin
+        # content; the charm is able to write that.
+        if self._stdin:
+            raise ValueError(
+                "processes cannot start with existing stdin content - write to "
+                "the process in your charm",
+            )
+
     def _run(self) -> int:
         return self._change_id
 
+    def _update_stdin(self, stdin: str):
+        # bypass frozen dataclass
+        object.__setattr__(self, "_stdin", stdin)
 
-_ExecMock = Dict[Tuple[str, ...], ExecOutput]
+    @property
+    def stdin(self):
+        """The content the charm wrote to the mock process's stdin."""
+        return self._stdin
```

Comment:
Perhaps, stdin is always string-like (str,bytes), and can never be None?

Or is None used as a sentinel?

🤔 I wonder if perhaps this should raise if the given Exec was never called... wdyt? Or... does Scenario only instantiate this object when charm attempts to exec something?

Reply:
> Perhaps, stdin is always string-like (str,bytes), and can never be None?

I'm completely ignoring bytes at the moment, which is probably wrong (but also unchanged from earlier Scenario). If you consistently use bytes then it should work apart from static checks complaining, I think. It probably ought to be explicit, although it would be nice to avoid all the overloading that ops has (but maybe that's unavoidable given that this is mocking ops).

To your actual point: it's currently None when the object is created and then will be a string in the output state, as long as `wait()` or `wait_output()` was called on the process. Perhaps `""` would be a better default value, although if it does grow to handle bytes as well then that's more complex, so `None` is probably better in that case. But this might end up immaterial, see the long comment above.

> 🤔 I wonder if perhaps this should raise if the given Exec was never called... wdyt? Or... does Scenario only instantiate this object when charm attempts to exec something?

`Exec` is instantiated by charm tests, but `_ExecProcess` is only instantiated by Scenario when something gets exec'd.

I don't think raising if an `Exec` isn't used is a good idea:
* It doesn't really align with the general Scenario behaviour. The tests don't care if you have a secret or relation or network or whatever that isn't used.
* It's convenient to have a fixture or similar helper that gives state components, and to be able to use that for a bunch of tests, only some of which might be executing the processes.
* It doesn't really match the real world: the executable is (most likely) still in the filesystem regardless of whether you actually execute it or not.